heudiconv/bids.py

```diff
@@ -457,3 +461,145 @@ def convert_sid_bids(subject_id):
+    - for fmap files: use the <acq> entity from the file name
+    - for other files: use the folder name ('anat', 'func', 'dwi', ...)
+
+    Parameters:
+    ----------
+    json_file : str
+
+    Returns:
+    -------
+    str with "ShimSetting" value or <acq> entity
+    """
+    data = load_json(json_file)
+    if SHIM_KEY in data.keys():
+        shims = data[SHIM_KEY]
+    else:
+        modality = op.basename(op.dirname(json_file))
+        if modality == 'fmap':
+            # extract the <acq> entity:
+            shims = re.search('(?<=[/_]acq-)\w+',json_file).group(0).split('_')[0]
+            if shims.lower() in ['fmri', 'bold']:
+                shims = 'func'
+        else:
+            shims = modality
+    return shims
+
+
+def populate_intended_for(path_to_bids_session):
+    """
+    Adds the 'IntendedFor' field to the fmap .json files in a session folder.
+    It goes through the session folders and checks what runs have the same
+    'ShimSetting' as the fmaps. If there is no 'ShimSetting' field in the json
+    file, we'll use the folder name ('func', 'dwi', 'anat') and see which fmap
+    with a matching '_acq' entity.
+
+    If several fmap runs have the same 'ShimSetting' (or '_acq'), it will use
+    the first one. Because fmaps come in groups (with reversed PE polarity,
+    or magnitude/phase), it adds the same runs to the 'IntendedFor' of the
+    corresponding fmaps by checking the '_acq' and '_run' entities.
+
+    Note: the logic behind the way we decide how to populate the "IntendedFor"
+    is: we want all images in the session (except for the fmap images
+    themselves) to have AT MOST one fmap.  (That is, a pair of SE EPI with
+    reversed polarities, or a magnitude a phase field map). If there are more
+    than one fmap (more than a fmap pair) with the same acquisition parameters
+    as, say, a functional run, we will just assign that run to the FIRST pair,
+    while leaving the other fmap pairs without any assigned images.  If the
+    user's intentions were different, he/she will have to manually edit the
+    fmap json files.
+
+    Parameters:
+    ----------
+    path_to_bids_session : str or os.path
+        path to the session folder (or to the subject folder, if there are no
+        sessions).
+    """
+    lgr.info('')
+    lgr.info('Adding "IntendedFor" to the fieldmaps in {}.'.format(path_to_bids_session))
+
+    # Resolve path (eliminate '..')
+    path_to_bids_session = op.abspath(path_to_bids_session)
+
+    # get the BIDS folder (if "data_folder" includes the session, remove it):
+    if op.basename(path_to_bids_session).startswith('ses-'):
+        bids_folder = op.dirname(path_to_bids_session)
+    else:
+        bids_folder = path_to_bids_session
+
+    fmap_dir = op.join(path_to_bids_session, 'fmap')
+    if not op.exists(fmap_dir):
+        lgr.warning('Fmap folder not found in {}.'.format(path_to_bids_session))
+        lgr.warning('We cannot add the IntendedFor field')
```

Comment:
> NB generally `%` form + passing positional args as args to the call is preferred for `lgr` since it would not even cause strings interpolation unless logging is enabled for the level.

Thanks for the tip!

Addressed in 9b397de36544e297b88cf27983bcbfb05f0d7b9a